### PR TITLE
kfctl: add --config flag for specifying a static config file

### PR DESCRIPTION
**Pull request by yanniszark**
As per the discussion in #3419, it seems that we want to move to an approach where:
* Each platform specifies its own set of static yamls files for the KfDef config.
* A user can specify a KfDef config to use, which should give us reproducibility, easy sharing of configs and is in-line with the GitOps philosophy.

This PR adds the `--config` flag to `kfctl`.
When set, `kfctl` will read the KfDef from the specified file instead of generating it using kustomize.
Platforms can also override this option and specify their own static file.

Partially related to: #2831 support pulling from a local repository
   * This will pull the config from a local repository but we still have to support adding and pulling
      all of the repos/manifests on which we depend from a local repo.

Signed-off-by: Yannis Zarkadas <yanniszark@arrikto.com>

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/3468)

### CODE REVIEW

File: bootstrap/pkg/kfapp/coordinator/coordinator.go [hunk 1]
```diff
@@ -63,9 +63,28 @@ func GetKfApp(kfdef *kfdefsv2.KfDef) kftypes.KfApp {
 	return _coordinator
 }
 
-func getConfigFromCache(pathDir string, kfDef *kfdefsv2.KfDef) ([]byte, error) {
+func getStaticConfigForPlatform(kfDef *kfdefsv2.KfDef) string {
```
Comment (jlewi) [inline bootstrap/pkg/kfapp/coordinator/coordinator.go]: What's the purpose of this function? Shouldn't any config  go into the static file?

File: bootstrap/pkg/kfapp/coordinator/coordinator.go [hunk 2]
```diff
@@ -63,9 +63,28 @@ func GetKfApp(kfdef *kfdefsv2.KfDef) kftypes.KfApp {
...
+		if err != nil {
+			return nil, &kfapis.KfError{
+				Code:    int(kfapis.INTERNAL_ERROR),
+				Message: fmt.Sprintf("error reading specified config file %s: %v", staticConfigFile, err),
+			}
+		}
+		return configFileBytes, nil
+	}
+
+	configPath := filepath.Join(kfDef.Spec.Repo, kftypes.DefaultConfigDir)
```
Comment (jlewi) [inline bootstrap/pkg/kfapp/coordinator/coordinator.go]: It looks like we are looking at a KfDef spec in order to locate a YAML file from which we get the KfDef spec.

Can we simplify this and not use a KfDef spec to provide whatever info (e.g. repo information to locate the yaml file).

For example suppose the syntax was 

```
kfctl init --config=file://myrepo/path/kfdef.yaml
```

Then the argument fully specifies the location so there's no reason to reference some other kfdef spec to figure out where to locate it from.

**Comment by jlewi** [comment]
I was expecting to change the code here
https://github.com/kubeflow/kubeflow/blob/7ca3cfb39368eb49042992f4e16215e41e1953c6/bootstrap/pkg/kfapp/coordinator/coordinator.go#L258

The current code is a bit roundabout. 

```
	kfDef := &kfdefsv2.KfDef{
...
	}
	configFileBuffer, configFileErr := getConfigFromCache(cacheDir, kfDef)
```

Its roundabout because we are creating a partially constructed `KfDef` and then using that to then read a `KfDef` spec from a yaml file. This leads to confusion about what happens if the original kfDef spec and the one in the yaml file have conflicts. 

Why not just do something like the following

````
pathToConfigFile := cacheDir + "/" + args.config
kfDef, err := loadKfDef(pathToConfigFile)
````

File: bootstrap/pkg/kfapp/coordinator/coordinator.go (hunk 1 above)
Comment (yanniszark) [reply bootstrap/pkg/kfapp/coordinator/coordinator.go]: The idea is:
* If the user has specified a config file, that file is used.
* If the user hasn't specified a config file, the config value can be seeded with a static config file for the specific platform.

File: bootstrap/pkg/kfapp/coordinator/coordinator.go (hunk 2 above)
Comment (yanniszark) [reply bootstrap/pkg/kfapp/coordinator/coordinator.go]: I agree that it's a convoluted way to do it and that we should refactor this part of kfctl to be more simple/clear in what it's doing.
However, I  believe it is a bit out of scope for this PR and would be a pretty big PR on its own.

**Comment by yanniszark** [comment]
@jlewi indeed the way we construct KfDef is convoluted and we need to make it more clear and simple.
Right now, after the KfDef is constructed with kustomize, the following values are overridden:
```go
	kfDef.Name = appName
	kfDef.Spec.AppDir = appDir
	kfDef.Spec.Platform = platform
	kfDef.Namespace = namespace
	kfDef.Spec.Version = version
	kfDef.Spec.Repo = path.Join(cacheDir, kftypes.KubeflowRepo)
	kfDef.Spec.Project = options[string(kftypes.PROJECT)].(string)
	kfDef.Spec.SkipInitProject = options[string(kftypes.SKIP_INIT_GCP_PROJECT)].(bool)
	kfDef.Spec.UseBasicAuth = useBasicAuth
	kfDef.Spec.UseIstio = useIstio
	kfDef.Spec.PackageManager = packageManager
```

In the case of `--config` should these be also overriden?
Should they be left as they are in the config file?
I believe this is something that merits further discussion.

For this PR, I made --config act as a replacement to the kustomize process of creating the config file.
I think that is a good place to start.

PS. I also added a small library (go-openuri) so that --config can easily accept either local files or http(s) schemes.

File: bootstrap/pkg/kfapp/coordinator/coordinator.go (hunk 1 above)
Comment (jlewi) [reply bootstrap/pkg/kfapp/coordinator/coordinator.go]: I'm not sure I follow. Isn't the idea to get rid of the paltform flag and replace it with a static config file?

i.e. instead of specifying --platform=gcp, I'd specify 
```
--config=kfctl_gcp_istio.yaml
```

To put it another way. What if we made it an error to specify both "--platform" and "--config"? Would we still need this function?

**Comment by jlewi** [comment]
You raise a good point about keeping the delta in this PR small and therefore continuing to override certain values in KfDefSpec after loading KfDefSpec.

What does that have to though with constructing a partially defined KfDefSpec in order to locate the KfDefSpec.yaml we want to load?

Can't the steps be

1. Look at --config
1. If --config isn't a full URI; append defaults as needed (e.g. assume its in kubeflow/kubeflow on a particular branch)
1. Load KfDefSpec.yaml - just call your go-get-URI function.
1. At this point processing KfDef should be the same regardless of whether it was produced via static config file or via kustomize.
1. Override KFDefSpec with any values as needed

Am I missing something?

**Comment by yanniszark** [comment]
Thanks for the review @jlewi! I agree with many concerns that you raise. Let me elaborate:

> What does that have to though with constructing a partially defined KfDefSpec in order to locate the KfDefSpec.yaml we want to load?

The `getConfigFromCache` function, which exists in kfctl before this PR, uses a partially constructed KfDef to find the correct paths in order to apply kustomization and generate the final KfDef.
The `getStaticConfigForPlatform` function that this PR introduces just follows the example of that function.
However, I agree it would be more correct to just pass whatever arguments are necessary.
I will change the function arguments.

The steps right now are:
1. Load --config value. If a value isn't given by the user, check if the platform wants to specify a static config file. In other words, `--platform=existing` becomes synonymous with `--config=.cache/kubeflow/kubeflow/master/bootstrap/config/platform_existing.yaml`.
2. If --config is set, load it. Otherwise generate it with kustomization and load it.
3. Override certain values.

**Comment by jlewi** [comment]
Thanks @yanniszark Let me know when this is ready for another look.

**Comment by jlewi** [comment]
Thinking longer term does this confusion go away if  [repo](https://github.com/kubeflow/kubeflow/blob/d491a71e14284cb052f84004fc36975d59beced2/bootstrap/config/base/kfctl_default.yaml#L6) property that's in KFDef points to the proper git location for that file?

For example, we will have a static config `kfctl_gcp_istio.yaml` on the v0.6-branch. The value of repo in that config file should point to the repos on that branch e.g. https://github.com/kubeflow/kubeflow/blob/v0.6-branch/

So kfctl takes a single parameter pointing at the static file containing KfDef and then reads the repos from that file.

I think we should remove from kfctl the overriding of the repos field in KfDef and probably other fields. I don't think we need it anymore and its causing confusion.

In other words, I think the desired semantics should be similar to how kubectl works;

e.g with kubectl you can create a YAML file

```
kubectl create -f myresource.yaml
```

If you want to customize the resource you do it by editing the yaml file before you run create. You don't generally supply additional command line arguments to the create function to override certain parts of  the spec e.g. you don't do

```
kubectl create -f myresource.yaml --name=myresource
```

**Comment by yanniszark** [comment]
@jlewi I really like this approach and the similarities to kubectl.
This also makes init similar to generate and apply, in that they also load a KfDef file, app.yaml.

How do you suggest we go about this for 0.6?
Keep both or aim for having --config and deprecating --platform and others?

**Comment by jlewi** [comment]
@yanniszark Lets try to get this PR in ASAP because once we have config taking in a complete YAML file we can avoid using overlays which I think will improve velocity a lot.

For this PR I think we just want to make the minimal changes to

1. Take in --config
1. If --config is not a complete URI make it a complete URI based on some sensible default; i.e. kubeflow/kubeflow@master
1. Load that via gooopenURI

**Comment by yanniszark** [comment]
@jlewi got it!
I'll have the first iteration asap.

**Comment by jlewi** [comment]
Thanks @yanniszark

File: bootstrap/cmd/kfctl/cmd/init.go [hunk 3]
```diff
@@ -61,6 +61,7 @@ or a <name>. If just <name>, a directory <name> will be created in the current d
 		}
 
 		useIstio := initCfg.GetBool(string(kftypes.USE_ISTIO))
```
Comment (kkasravi) [inline bootstrap/cmd/kfctl/cmd/init.go]: wouldn't we want to remove all other cli options in Init like --project, --platform, --namespace, --repo, ...

File: bootstrap/pkg/kfapp/coordinator/coordinator.go (hunk 1 above)
Comment (kkasravi) [reply bootstrap/pkg/kfapp/coordinator/coordinator.go]: i think command completion should provide a list of static config files available to help the user
eg
--config kfctl_<tab><tab> kfctl_default.yaml kfctl_default_istio.yaml kfctl_gcp.yaml kfctl_gcp_istio.yaml, ...

File: bootstrap/cmd/kfctl/cmd/init.go (hunk 3 above)
Comment (yanniszark) [reply bootstrap/cmd/kfctl/cmd/init.go]: I agree that the plan is to remove all of those flags.
However, as this is PR should not break anything, it will just create a different code path.
Once all platforms migrate, we should be able to delete the old code path.

**Comment by yanniszark** [comment]
@jlewi the first iteration is ready.

* The `--config` flag accepts either a path or a URL. If the protocol is missing in front, it is assumed to be a file.
* We still need to override some fields to maintain the same functionality. We need to override these because their values are mainly determined at runtime. Namely:
    * `repo`: contains the information of where the kubeflow cached repo resides. 
    * `appDir`: container the information of where the app.yaml resides.

After deciding on a better practice for `repo` in #3471, we should be able to make this procedure cleaner.

**Comment by jlewi** [comment]
/lgtm
/approve

/hold 

I only had 2 nitpicky comments otherwise this looks great.
